launch_ros/launch_ros/actions/lifecycle_node.py

```diff
@@ -0,0 +1,123 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for the LifecycleNode action."""
+
+import functools
+import logging
+from typing import cast
+from typing import List
+from typing import Optional
+from typing import Text
+
+import launch
+from launch.action import Action
+
+import lifecycle_msgs.msg
+import lifecycle_msgs.srv
+
+from .node import Node
+from ..events.lifecycle import ChangeState
+from ..events.lifecycle import StateTransition
+
+_logger = logging.getLogger(name='launch_ros')
+
+
+class LifecycleNode(Node):
+    """Action that begins executing a process and sets up event handlers for the process."""
+
+    def __init__(self, *, node_name: Text, **kwargs) -> None:
+        """
+        Construct an LifecycleNode action.
+
+        Almost all of the arguments are passed to :class:`Node` and eventually
+        to :class:`launch.actions.ExecuteProcess`, so see the documentation of
+        those classes for additional details.
+
+        This action additionally emits some event(s) in certain circumstances:
+
+        - :class:`launch.events.lifecycle.StateTransition`:
+
+            - this event is emitted when a message is published to the
+              "/<node_name>/transition_event" topic, indicating the lifecycle
+              node represented by this action changed state
+
+        This action also handels some events related to lifecycle:
+
+        - :class:`launch.events.lifecycle.ChangeState`
+
+          - this event can be targeted to a single lifecycle node, or more than
+            one, or even all lifecycle nodes, and it causes the targeted nodes
+            to change state if possible, see its documentation for more details.
+        """
+        super().__init__(node_name=node_name, **kwargs)
+        self.__rclpy_subscription = None
+        # first item is unknown state
+        self.__current_state = next(iter(ChangeState.valid_states.values()))
+
+    def _on_transition_event(self, context, msg):
+        try:
+            event = StateTransition(action=self, msg=msg)
+            self.__current_state = ChangeState.valid_states[msg.goal_state.id]
+            context.emit_event_sync(event)
+        except Exception as exc:
+            print("Exception in handling of 'lifecycle.msg.TransitionEvent': {}".format(exc))
+
+    def _call_change_state(self, request, context: launch.LaunchContext):
+        while not self.__rclpy_change_state_client.wait_for_service(timeout_sec=1.0):
+            if context.is_shutdown:
+                _logger.warn("Abandoning wait for the '{}' service.".format(
+                    self.__rclpy_change_state_client.srv_name))
+                return
+        response = self.__rclpy_change_state_client.call(request)
+        if not response.success:
+            _logger.warn("Failed to make transition '{}' for LifecycleNode '{}'".format(
```

Comment:
maybe this should be an error instead of just a warning.

Reply:
65c2b53